### PR TITLE
Create Currency Modal, with geolocation and Location permissions

### LINKED ISSUE #1970: [Pay on Thursday 27th May] [IOU] Create Currency Modal, with geolocation and Location permissions
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!
___

## Context

This issue builds on top of the existing IOU Modal, which was implemented [in this PR](https://github.com/Expensify/Expensify.cash/pull/1561). The `IOU Amount` step page has already [been created](https://github.com/Expensify/Expensify.cash/blob/master/src/pages/iou/steps/IOUAmountPage.js) and is a controlled component ([IOUModal](https://github.com/Expensify/Expensify.cash/blob/fa6185b2c51fdc0f28ab9cbc48b8a306f64a537d/src/pages/iou/IOUModal.js#L194-L198) manages its state). The amount step allows users to select the IOU request or bill split amount, depending on how they launched the Modal. 

We now seek to extend functionality. We want to use a Geolocation library and an existing API to detect the users preferred currency, allowing them to change it manually, too. We also need to implement a permissions library to request location permission from the user if they decide to change their currency, as we will then seek to detect it automatically from then on.

To open the Modal, you can navigate to `localhost:8080/iou/request` AND `localhost:8080/iou/split`.

You can also launch the IOUModal by temporarily modifying the routes in CreateMenu. Launch the request money flow with `Navigation.navigate(ROUTES.IOU_REQUEST)` and the bill split flow with `Navigation.navigate(ROUTES.IOU_BILL)`

**Mobile & mobile web**
Currency Selector - Mobile

**Web & Desktop**
Currency Selector - Desktop (1)

**Deliverables:**

1. Create a new API.js function for calling our `GetPreferredCurrency` API and another for calling `GetCurrencyList`. Details at the bottom of this post.
2. Add the [react-native-geolocation](https://github.com/react-native-geolocation/react-native-geolocation#getcurrentposition) and [react-native-permissions](https://github.com/zoontek/react-native-permissions) libraries to the project.
3. We should create a new function within the [PersonalDetails.js Action](https://github.com/Expensify/Expensify.cash/blob/master/src/libs/actions/PersonalDetails.js) which calls the new `GetPreferredCurrency` API.js function
    1. If the location permission has already been granted, we will get lat and lon from getCurrentPosition and pass the params to `GetPreferredCurrency`, else we ignore those query params (as they are optional)
    2. We should then call `GetCurrencyList` matching the currency we received from the previous request to get the correct symbol for that currency. (the full currency list should also be persisted to Onyx via a new key: `currencyList`)
    3. We then persist both the `preferredCurrencyCode` AND `preferredCurrencySymbol` data in Onyx, similar to [how we store the users timezone](https://github.com/Expensify/Expensify.cash/blob/dbb40b9ea4358ea8a0d192a72d1fd232169bcc8c/src/libs/actions/PersonalDetails.js#L113-L114)
3. When the IOUModal is opened, we should subscribe to the `MY_PERSONAL_DETAILS` Onyx key:
    1. We take updated values of `preferredCurrencySymbol` and `preferredCurrencyCode`, setting IOUModal props for both of these (`currencySelected` prop should be deleted,  `selectedCurrency` should be modified to be of type Object: `selectedCurrency: { currencyCode: String, currencySymbol: String }`)
    2. The currency Text component should be updated with `this.props.selectedCurrency.currencySymbol` 
4. When the currency symbol is tapped, a CurrencyModal should be displayed. This is a new Modal which needs the following functionality:
    1.  The Modal should follow the structure of our existing Modals
    2. Retrieve all currencies from `GetCurrencyList` via another new Action function and again persisting this data to Onyx under the `currencyList` key.
    3. using the above data, we should display all currencies to the user in a FlatList, using a new [OptionsListUtils](https://github.com/Expensify/Expensify.cash/blob/master/src/libs/OptionsListUtils.js) variant 
    4. If the user makes a currency selection then we should pass this value back to the IOUModal via a callback and update the `preferredCurrencyCode` AND `preferredCurrencySymbol` Onyx data
    5. The modal should then close, returning the user to the IOUAmount step, with the updated currency symbol

**GetPreferredCurrency API**
Example of the API which is used to retrieve the users preferred currency
```
https://www.expensify.com/api?command=GetPreferredCurrency&latitude=0.01&longitude=-1.44
latitude - optional param
longitude - optional param

// Response
{
  "currency": "GBP",
  "jsonCode": 200,
  "requestID": "0uBdrX"
}
```

**GetCurrencyList API**
Example of the API which returns all supported currency and symbol parings
```
https://www.expensify.com/api?command=Mobile_GetConstants&data=[%22currencyList%22]

// Response
{
  "currencyList": "{"AED":{"symbol":"Dhs","name":"UAE Dirham","ISO4217":"784"},"AFN":{"symbol":"Af","name":"Afghan Afghani","ISO4217":"971"},"ALL":{"symbol":"ALL","name":"Albanian
	...
  }}",
  "jsonCode": 200,
  "requestID": "wjy42x"
}
```

## Upwork Issue: https://www.upwork.com/ab/applicants/1373765458169090048/job-details

### PR DESCRIPTION
**Pull request by rameshhpathak**
<If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit.>

### Details
<Explanation of the change or anything fishy that is going on>

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes #1970

### Tests

1. Go to /iou/split on web. Verify that the currency is based on your location.
2. Click on the currency, a list with all currencies should show.
3. Search for a currency, or scroll down to your desired currency 
4. Click on the currency to select and click on the Confirm button the confirm selection
5. Verify that the selected currency shows now in the amount page
6. Refresh the page, and verify that the same currency shows in the amount page

### Tested On

- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [x] iOS
- [x] Android

#### Web

Screenshot 2021-03-27 at 15 03 41
Screenshot 2021-03-27 at 15 05 30
Screenshot 2021-03-27 at 15 05 39
Screenshot 2021-03-27 at 15 05 47

#### iOS

https://user-images.githubusercontent.com/30867740/112727445-a0058400-8f4a-11eb-89ff-efbddf283607.mov

#### Android

Screenshot 2021-03-27 at 15 28 28
Screenshot 2021-03-27 at 15 28 39
Screenshot 2021-03-27 at 15 28 44
Screenshot 2021-03-27 at 15 28 51

### CODE REVIEW

**Comment by rameshhpathak** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by rameshhpathak** [comment]
recheck

**Comment by rameshhpathak** [comment]
@Julesssss  @iwiznia any idea on why this is not accessing react-native-onyx? It works OK on local environment

File: src/pages/iou/IOUCurrencySelection.js [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    render() {
+        const sections = this.getSections();
+        return (
+            <ScreenWrapper>
+                {() => (
+                    <>
+                        <View style={[styles.flex1, styles.w100]}>
+                            <View style={[styles.ph5, styles.pv3]}>
```
Comment (shawnborton) [inline src/pages/iou/IOUCurrencySelection.js:161]: Since this entire area is already wrapped in 20px padding, let's remove the padding top from this view.

**Comment by shawnborton** [comment]
Maybe this is just a matter of pulling master, but by default, our modal views should not have a bottom border under the title. So the border you see here should not be included:
image


CC @Julesssss in case maybe this needs to be changed elsewhere.

**Comment by Julesssss** [comment]
> Maybe this is just a matter of pulling master, but by default, our modal views should not have a bottom border under the title. So the border you see here should not be included:

Yeah, this change will need to be made elsewhere. I'll create an issue for it.

**Comment by iwiznia** [comment]
> @Julesssss @iwiznia any idea on why this is not accessing react-native-onyx? It works OK on local environment

What do you mean exactly? What's not working? I see you are using Onyx already...

**Comment by rameshhpathak** [comment]
> > @Julesssss @iwiznia any idea on why this is not accessing react-native-onyx? It works OK on local environment
> 
> What do you mean exactly? What's not working? I see you are using Onyx already...

The checks are failing with this error after not able to access onyx. 
onyx

**Comment by Julesssss** [comment]
Hi @rameshhpathak, 

You can resolve this by running `pod install` locally from the iso folder, then committing the changed files (`ios/Podfile.lock` & `package-lock.json`). Then, due to a conflict, you'll need to merge the latest changes from master. 

I confirmed this on a test PR here: https://github.com/Expensify/Expensify.cash/pull/2152

**Comment by Julesssss** [comment]
Hi @rameshhpathak, 

I'm really not sure why this happened, but I was able to solve this by [making your commit in my name](https://github.com/Expensify/Expensify.cash/pull/2153). You can see here that the tests are failing for JSLint errors now, instead of the auth error. 

I would suggest that you try cherry-picking your changes onto a clean branch/PR to see if that solves the issue.

**Comment by Julesssss** [comment]
Maybe you know this already, but the Podfile wasn't updated either, so you'll need to run pod install from /ios

**Comment by rameshhpathak** [comment]
> Hi @rameshhpathak,
> 
> I'm really not sure why this happened, but I was able to solve this by [making your commit in my name](https://github.com/Expensify/Expensify.cash/pull/2153). You can see here that the tests are failing for JSLint errors now, instead of the auth error.
> 
> I would suggest that you try cherry-picking your changes onto a clean branch/PR to see if that solves the issue (this is probably the quickest path forward)

I'll do that. Thank you for your help.

**Comment by Julesssss** [comment]
Hey @rameshhpathak, any luck with the suggested fix?

**Comment by rameshhpathak** [comment]
> Hey @rameshhpathak, any luck with the suggested fix?

Hey @Julesssss, I couldn't find time to do it yesterday. Been trying - failed two times, third time lucky perhaps?

**Comment by rameshhpathak** [comment]
@Julesssss  I installed it fresh in this PR #2194 . WIll close this PR now :) Thank you for your help!!!

**Pull request closed without merging**